Reluctantly restore compatibility of xsbti.*

eed3si9n opened the pull request:
Fixes https://github.com/sbt/zinc/issues/779

As it stands, compiler-interface and compiler bridge implementation are of internal concern of Zinc implementation. We _should_ be able to remove methods or change the method signatures. The word "interface" here is between Scalac and Zinc internal, not to the world.

In reality, the situation is more complicated because we have Dotty compiler out there that is bound to specific version of compiler-interface. So when I released sbt 1.4.0-M1 this resulted in NoSuchMethodErrors:

```
[error] ## Exception when compiling 1 sources to /private/tmp/hello-dotty/target/scala-0.24/classes
[error] java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
[error] xsbt.CompilerInterface.newCompiler(CompilerInterface.java:35)
....
[error] Caused by: java.lang.NoSuchMethodError: xsbti.compile.SingleOutput.getOutputDirectory()Ljava/io/File;
[error]   at xsbt.CachedCompilerImpl.<init>(CachedCompilerImpl.java:35)
```

To smooth things out, one approach we've discussed is to create a separate compiler-interface (in a different package) that is less dependent on Zinc specifics. Related to that, in https://github.com/sbt/zinc/pull/661 I've created Java interface for `CompilerInterface1`, and we can use pattern matching to see which capability the compiler bridge implementation supports. This PR brings in the Java interfaces proposed in #661 + scala/compiler-interface.

In any case, this commit brings back the old `java.io.File`-based methods, and locally I was able to get hello world from Dotty.

eed3si9n committed:
Reluctantly restore compatibility of xsbti.*

Fixes https://github.com/sbt/zinc/issues/779

As it stands, compiler-interface and compiler bridge implementation are of internal concern of Zinc implementation. We _should_ be able to remove method or change the signatures. The word "interface" here is between Scalac and Zinc internal, not to the world.

In reality, the situation is more complicated because we have Dotty compiler out there that is bound to a specific version of compiler-interface. So when I released sbt 1.4.0-M1 this resulted in NoSuchMethodErrors:

```
[error] ## Exception when compiling 1 sources to /private/tmp/hello-dotty/target/scala-0.24/classes
[error] java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
[error] xsbt.CompilerInterface.newCompiler(CompilerInterface.java:35)
....
[error] Caused by: java.lang.NoSuchMethodError: xsbti.compile.SingleOutput.getOutputDirectory()Ljava/io/File;
[error]   at xsbt.CachedCompilerImpl.<init>(CachedCompilerImpl.java:35)
```

To smooth things out, one approach we've discussed is to create a separate compiler-interface (in a different package) that is less dependent on Zinc specifics. Related to that, in https://github.com/sbt/zinc/pull/661 I've created Java interface for `CompilerInterface1`, and we can use pattern matching to see which capability the compiler bridge implementation supports. This commit brings in those Java interfaces as well.

In any case, this commit brings back the old `java.io.File`-based methods, and locally I was able to get hello world from Dotty.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/829/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

dwijnand commented on internal/compiler-bridge/src/main/scala/xsbt/CompilerInterface.scala:
```diff
@@ -115,7 +115,17 @@ private final class CachedCompiler0(args: Array[String], output: Output, initial
+      log: Logger,
+      delegate: Reporter,
+      progress: CompileProgress
+  ): Unit = ???
```
How does this work for compatibility?  Should this be calling the other run somehow?

eed3si9n replied:
Zinc 1.4.x should never use this method to enter into this (source-compiled) bridge. If we ever reach in here that means there's a bug in pattern match behavior, which I want to fail spectacularly.

dwijnand replied:
Could you capture that in the thrown exception, please?  If there's more to say then appropriate for the message, then as a comment.  Thanks!

dwijnand replied:
Why does `CachedCompiler2` extends `CachedCompiler1` then?

eed3si9n replied:
Sure. I'll make those changes.

eed3si9n replied:
Ideally I would split `CachedCompiler2` and `CachedCompiler` but I think it's kind of annoying for that one because compiler cache is being used by sbt:

```scala
      compilerCache := state.value get Keys.stateCompilerCache getOrElse CompilerCache.fresh,
```

and the implementation is tied to `CachedCompiler` in https://github.com/sbt/zinc/blob/v1.4.0-M6/internal/zinc-compile-core/src/main/java/xsbti/compile/CompilerCache.java  + https://github.com/sbt/zinc/blob/v1.4.0-M6/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/CompilerCache.scala.

So in `CachedCompiler2` would have two `run` method. One that works, and the other that would throw an error.

eed3si9n replied:
In https://github.com/sbt/zinc/pull/829/commits/4774e53fb157b9aee24dce605897ebf417d543a7 I moved the error throwing to `CachedCompiler2#run`'s default method.

dwijnand commented on internal/compiler-bridge/src/main/scala/xsbt/CompilerInterface.scala:
```diff
@@ -19,16 +19,16 @@ import scala.reflect.io.AbstractFile
 
-final class CompilerInterface {
-  def newCompiler(
+final class CompilerInterface extends CompilerInterface2 {
```
Compared to the other parts of the diff, this seems backwards.  Why does CompilerInterface(1) extends CompilerInterface2?

eed3si9n replied:
`xsbt.CompilerInterface` is a hard-coded class name all bridges use as entry point.
`xsbti.CompilerInterface2` is the new Java interface.

smarter replied:
> xsbt.CompilerInterface is a hard-coded class name all bridges use as entry point

Now that we're coding against interfaces, we could replace the use of hardcoded class names by a file in META-INF listing the entry points of the bridge (cf https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html), maybe something worth considering for a follow-up PR?

eed3si9n replied:
> Now that we're coding against interfaces, we could replace the use of hardcoded class names by a file in META-INF listing the entry points of the bridge

Thanks for the pointer. That sounds like a good idea indeed.

dwijnand replied:
Ah, I see.  `CompilerInterface` was/is the source-based interface to the compiler, but it's actually implemented as a class (not an `interface`).

The new interface being defined are java `interface`s, with `CompilerInterface1` and `CompilerInterface2`, and the implementation here now implements `CompilerInterface2`.

Seeing as we're not dropping support for the older interface (the older methods, that were defined directly in this class), now captured by `CompilerInterface1`, I think it would be clearer for this to extend `CompilerInterface1` too.

dwijnand replied:
> Seeing as we're not dropping support for the older interface (the older methods, that were defined directly in this class), now captured by `CompilerInterface1`, I think it would be clearer for this to extend `CompilerInterface1` too.

Scratch that.  We don't drop support for _consuming_ the old interface (CompilerInterface1) which will continue to be supplied by current sbt-dotty (until they switch).  But the Scala 2 bridges by Zinc 1.4+ will be implementing the new, CompilerInterface2, methods.

dwijnand commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/CompilerCache.scala:
```diff
@@ -57,6 +58,17 @@ final class CompilerCache(val maxInstances: Int) extends GlobalsCache {
           override def commandArguments(sources: Array[File]): Array[String] = {
             compiler.commandArguments(sources)
           }
+          @silent
```
What's being silenced?

eed3si9n replied:
Deprecation warning that I'm using `run` that takes `Array[File]`.

dwijnand commented on build.sbt:
```diff
@@ -34,6 +34,7 @@ def mimaSettings: Seq[Setting[_]] = Seq(
   },
 )
 
+Global / onChangedBuildSource := ReloadOnSourceChanges
```
Could we not add this, please?  It's a departure from the default behaviour, which I prefer.

eatkins replied:
I agree that this setting is ideally either in `~/.sbt/global.sbt` or in a gitignored local file (I added `local.sbt` to my global gitignore for things like this).

eed3si9n replied:
I've been on a vim + `~` kick lately, and with `~` `ReloadOnSourceChanges` is really convenient. If you guys think it should be moved to `local.sbt` I'll do it, but shouldn't we be dogfooding all the knobs so we'll catch when it bugs out?

eatkins replied:
Is there a way that we could make the value in `~/.sbt/global.sbt` always win?

I do think that for many, but not all builds, auto reload is a much better default. I leave it off in sbt itself though. The reason is say that I bump a library version that is backward compatible with the old version, I then rebase to an earlier commit that was originally written against the old library api, I really don't want to reload both because reload itself takes a while in sbt (like 5 seconds or more -- much of which I now know is spent building the compiled setting map: https://github.com/sbt/sbt/pull/5666) and because changing the library might end up effectively invalidating the entire sbt project since zinc doesn't do api hashing of jars on the classpath and invalidates everything if the jar changes. If reload were fast and zinc were better (or configured differently) about invalidating binary dependencies, I think auto reload would be the no-brainer default.

eatkins replied:
I also agree with your point about dogfooding. https://github.com/sbt/sbt/pull/5673/commits/366c49a7641de33e5520f8d0ec0898e0bb19b99e was a direct result of my dogfooding the vim + `~` workflow in a new project I'm working on.

eed3si9n replied:
Here's my proposal related to this - https://github.com/sbt/sbt/issues/5679

eed3si9n replied:
https://github.com/sbt/zinc/pull/829/commits/8de29c50bd3a2e981092371c2b1972731723c05e no longer contains `Global / onChangedBuildSource`

dwijnand replied:
I would love to not have to think about whether I need to rebuild the state or not, but while it takes the time it does and how it affects zinc, I prefer to stay in control.

I'm really excited about the compiled setting map caching, hoping it doesn't hit any bugs.  I also built https://github.com/dwijnand/sbt-reloadquick/ way back when, so I could skip some waiting - perhaps there's something there that can be folded into `reload`.

dwijnand reviewed (commented):
Thanks, Eugene.  Looks reasonable to me, but I had a few questions:

smarter commented:
Thanks a lot for this PR! That's one less thing for me to worry about on the road to Scala 3 :).

eed3si9n committed:
Address review

dwijnand commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala:
```diff
@@ -177,12 +179,19 @@ final class AnalyzingCompiler(
+        intf.run(sources.toArray, arguments.toArray[String], log, reporter)
+    }
+    /*
+      call("xsbt.ScaladocInterface", "run", log)(
```
Need this to continue to support current sbt-dotty, no?

dwijnand replied:
Looks like this whole code is dead.

eed3si9n replied:
Yea. Looks like it's missing the fallback.

dwijnand commented on internal/compiler-interface/src/main/java/xsbti/compile/CachedCompiler2.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public interface CachedCompiler2 extends CachedCompiler {
+  default void run(File[] sources, DependencyChanges changes, AnalysisCallback callback, Logger logger, Reporter delegate, CompileProgress progress) {
+    throw new RuntimeException("run method with Array[File] is not supported for this compiler bridge (see sbt/zinc#829)"); 
```
Going to try and send a change to avoid this and the `.asInstanceOf[CachedCompiler2]`.

dwijnand committed:
Avoid unsafe code around CachedCompiler2

Avoid:
  * throwing if the "wrong" `run` method on CachedCompiler2 is called
  * casting the CachedCompiler value returned by GlobalsCache to
    CachedCompiler2

The easiest/safest way to do this is by making CachedCompiler2 extend
CachedCompiler, as much as I'd prefer not to.  Fortunately implementing
the old methods in our bridge's CachedCompiler implementation
(CachedCompiler0) is trivial.

dwijnand commented:
> Ideally I would split `CachedCompiler2` and `CachedCompiler` but I think it's kind of annoying for that one because compiler cache is being used by sbt:
> 
> ```scala
>       compilerCache := state.value get Keys.stateCompilerCache getOrElse CompilerCache.fresh,
> ```
> 
> and the implementation is tied to `CachedCompiler` in

That's `GlobalsCache`, which is both produced and consumed by Zinc.  So we can keep `CachedCompiler` around, for sbt-dotty, and have a separate `CachedCompiler2` as the replacement in Zinc.

eed3si9n commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala:
```diff
@@ -125,74 +122,14 @@ final class AnalyzingCompiler(
-     */
-    ()
+  ): CachedCompiler2 = {
+    val (bridge: CompilerInterface2, _) = bridgeInstance(compilerBridgeClassName, log)
```
How do you know that the bridge instance supports `CompilerInterface2`?

dwijnand replied:
Ah.  I thought `CompilerInterface` was Zinc, not Zinc's Scala 2 bridge...

eed3si9n replied:
I tried to save as much as possible here - https://github.com/sbt/zinc/pull/829/commits/10d69d54b8ecfba66d0271713e81b53e51215e03

eed3si9n commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala:
```diff
@@ -125,74 +122,14 @@ final class AnalyzingCompiler(
-
-  def doc(
-      sources: Seq[VirtualFile],
-      classpath: Seq[VirtualFile],
-      outputDirectory: Path,
-      options: Seq[String],
-      maximumErrors: Int,
-      log: ManagedLogger
-  ): Unit = {
-    val reporter = new ManagedLoggedReporter(maximumErrors, log)
-    doc(sources, classpath, outputDirectory, options, log, reporter)
-  }
```
Why is doc being deleted?

dwijnand replied:
Unused.

eed3si9n replied:
Used by sbt I think - https://github.com/sbt/sbt/blob/develop/main-actions/src/main/scala/sbt/Doc.scala

eed3si9n commented on internal/compiler-interface/src/main/java/xsbti/compile/CachedCompilerProvider.java:
```diff
@@ -22,5 +22,5 @@
 
 	/** Return a new cached compiler from the usual compiler input. */
-	CachedCompiler newCachedCompiler(String[] arguments, Output output, Logger log, Reporter reporter);
+	CachedCompiler2 newCachedCompiler(String[] arguments, Output output, Logger log, Reporter reporter);
```
AnalysisCompiler should be able to return either Dotty CachedCompiler or Scala 2.x CachedCompiler2, so if you want to split them, this should return AnyRef.

eed3si9n commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/CompilerCache.scala:
```diff
@@ -31,7 +29,7 @@ import xsbti.VirtualFile
 
   /** Define a least-recently used cache indexed by a generated key. */
-  private[this] val cache = lru[CompilerKey, CachedCompiler](maxInstances)
+  private[this] val cache = lru[CompilerKey, CachedCompiler2](maxInstances)
```
Same here. This would be AnyRef.

eed3si9n commented:
What would be useful is scripted test emulating sbt-dotty to download a bridge binary.

eed3si9n commented:
Rebased to resolve VirtualFileUtil.scala conflict.

eed3si9n committed:
Fix doc fallback

dwijnand commented:
> I tried to save as much as possible here - [10d69d5](https://github.com/sbt/zinc/commit/10d69d54b8ecfba66d0271713e81b53e51215e03)

The point of my change was to make it less unsafe.  Making it `AnyRef` is even worst...

I'll rebase my approach and share it.

eed3si9n commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/CompilerCache.scala:
```diff
@@ -45,31 +46,51 @@ final class CompilerCache(val maxInstances: Int) extends GlobalsCache {
-            // Dont' close the underlying Global.
-          }
+        val compiler0 = c.newCachedCompiler(args, output, log, reporter)
+        val newCompiler = compiler0 match {
+          case compiler: CachedCompiler2 =>
+            new CachedCompiler2 with java.io.Closeable {
+              override def run(
+                  sources: Array[VirtualFile],
+                  changes: DependencyChanges,
+                  callback: AnalysisCallback,
+                  logger: xLogger,
+                  delegate: Reporter,
+                  progress: CompileProgress
+              ): Unit = {
+                compiler.run(sources, changes, callback, logger, delegate, progress)
+              }
+              override def close(): Unit = {
+                // Don't close the underlying Global.
+              }
+            }
+          case compiler: CachedCompiler =>
+            (new CachedCompiler with java.io.Closeable {
+              override def commandArguments(sources: Array[File]): Array[String] = {
+                compiler.commandArguments(sources)
+              }
+              override def run(
+                  sources: Array[File],
+                  changes: DependencyChanges,
+                  callback: AnalysisCallback,
+                  logger: xLogger,
+                  delegate: Reporter,
+                  progress: CompileProgress
+              ): Unit = {
+                compiler.run(sources, changes, callback, logger, delegate, progress)
+              }
+              override def close(): Unit = {
+                // Don't close the underlying Global.
+              }
+            }: @silent // silence deprecation of CachedCompiler
+            )
```
We can push this (wrapping to hijack `close()`) out to `AnalyzingCompiler` to make it a bit more clearer.

dwijnand replied:
No, because it's only cached instance we don't want to close.  The fresh ones it's best if they get closed out.

eed3si9n commented on internal/compiler-interface/src/main/java/xsbti/compile/CachedCompilerProvider.java:
```diff
@@ -22,5 +22,5 @@
 
 	/** Return a new cached compiler from the usual compiler input. */
-	CachedCompiler newCachedCompiler(String[] arguments, Output output, Logger log, Reporter reporter);
+	Object newCachedCompiler(String[] arguments, Output output, Logger log, Reporter reporter);
```
Maybe return `Optional<CachedCompiler>` and add `newCachedCompiler2` that returns `Optional<CachedCompiler2>`?

dwijnand replied:
That still leaves the possibility of both returning empty - then what?

eed3si9n replied:
Throw?

eed3si9n replied:
Either way, we can work on that in a follow-up PR to remove cached compiler in general.

eed3si9n commented:
Also maybe this is a good opportunity to remove `CompilerCache` in general.
It was turned off by default 

```scala
compilerCache <<= state map { _ get Keys.stateCompilerCache getOrElse compiler.CompilerCache.fresh },
```

in https://github.com/sbt/sbt/commit/6769c942083d37c7729adcd26f43d441ab89ecf1 when it was first introduced in 2012 . There was an attempt to remove it in 2017 https://github.com/sbt/zinc/pull/345 / https://github.com/sbt/sbt/issues/1287 / https://github.com/sbt/zinc/issues/340, but it didn't go all the way. So @retronym added a TODO to remove it in 2019 when the compiler became `java.io.Closable` in https://github.com/scala/scala/pull/7366 - https://github.com/sbt/zinc/pull/643.

eed3si9n commented:
ok. Let's merge if this one passes. Yea?
